File: paleo_workbench/ui/pages/map_edit_view.py
from __future__ import annotations

from PySide6.QtCore import Qt, QTimer, Signal
from PySide6.QtGui import QKeyEvent, QMouseEvent, QPainter, QWheelEvent
from PySide6.QtWidgets import QGraphicsView

from paleo_workbench.ui import tokens
from paleo_workbench.ui.pages.map_edit_scene import MapEditScene

# Idle delay before full-detail rendering is restored after navigation.
_NAV_LOD_IDLE_MS = 120


class MapEditView(QGraphicsView):
    """Primary map edit surface backed by MapEditScene."""

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setObjectName("MapEditView")
        self.setStyleSheet(
            f"QGraphicsView#MapEditView {{ background: {tokens.BG_SEARCH};"
            f" border: 1px solid {tokens.BORDER};"
            f" border-radius: {tokens.RADIUS_CARD}px; }}"
        )
        self.setRenderHint(QPainter.RenderHint.Antialiasing, True)
        self.setDragMode(QGraphicsView.DragMode.NoDrag)
        self.setTransformationAnchor(QGraphicsView.ViewportAnchor.AnchorUnderMouse)
        self.setResizeAnchor(QGraphicsView.ViewportAnchor.AnchorUnderMouse)
        self.setViewportUpdateMode(QGraphicsView.ViewportUpdateMode.SmartViewportUpdate)
        self.setFocusPolicy(Qt.FocusPolicy.StrongFocus)
        self.setMouseTracking(True)
        self._shared_view_state = {"center": (0.0, 0.0), "scale": 1.0}
        # Navigation display LOD state (display-only; restored after idle).
        self._nav_lod_active = False
        self._nav_lod_timer = QTimer(self)
        self._nav_lod_timer.setSingleShot(True)
        self._nav_lod_timer.setInterval(_NAV_LOD_IDLE_MS)
        self._nav_lod_timer.timeout.connect(self._end_navigation_lod)

        scene = MapEditScene(self)
        self.setScene(scene)

    # --- navigation display LOD ---------------------------------------------

    def navigation_lod_active(self) -> bool:
        return self._nav_lod_active

    def _begin_navigation_lod(self) -> None:
        """Enter low-detail mode for wheel/pan; an idle timer restores detail."""
        if not self._nav_lod_active:
            self._nav_lod_active = True
            self.setRenderHint(QPainter.RenderHint.Antialiasing, False)
            self.setOptimizationFlags(
                QGraphicsView.OptimizationFlag.DontAdjustForAntialiasing
                | QGraphicsView.OptimizationFlag.DontSavePainterState
            )
            scene = self.scene()
            if isinstance(scene, MapEditScene):
                scene.set_navigation_lod(True)
        self._nav_lod_timer.start()  # restart the idle countdown

    def _end_navigation_lod(self) -> None:
        if not self._nav_lod_active:
            return
        self._nav_lod_active = False
        self._nav_lod_timer.stop()
        self.setRenderHint(QPainter.RenderHint.Antialiasing, True)
        self.setOptimizationFlags(QGraphicsView.OptimizationFlag(0))
        scene = self.scene()
        if isinstance(scene, MapEditScene):
            scene.set_navigation_lod(False)
        self.viewport().update()

    # --- events ---------------------------------------------------------------

    def wheelEvent(self, event: QWheelEvent) -> None:
        """Zoom with mouse wheel (no geometry change); low-detail while scrolling."""
        delta = event.angleDelta().y()
        if delta == 0:
            event.ignore()
            return
        self._begin_navigation_lod()
        factor = 1.15 if delta > 0 else 1.0 / 1.15
        self.scale(factor, factor)
        self._shared_view_state = self._read_view_state()
        self.view_state_changed.emit(self.view_state())
        event.accept()

    def mouseMoveEvent(self, event: QMouseEvent) -> None:
        """Publish the cursor in scene coordinates (project CRS)."""
        pos = self.mapToScene(event.position().toPoint())
        self.cursor_position_changed.emit((float(pos.x()), float(pos.y())))
        super().mouseMoveEvent(event)

    def scrollContentsBy(self, dx: int, dy: int) -> None:
        # Any pan (scrollbars, keyboard, centerOn) engages navigation LOD.
        if dx or dy:
            self._begin_navigation_lod()
        super().scrollContentsBy(dx, dy)

    def _read_view_state(self) -> dict:
        center = self.mapToScene(self.viewport().rect().center())
        return {
            "center": (float(center.x()), float(center.y())),
            "scale": float(self.transform().m11()),
        }

    def view_state(self) -> dict:
        return dict(self._shared_view_state)

    def apply_view_state(self, state: dict, *, emit: bool = False) -> None:
        center = tuple(state.get("center", (0.0, 0.0)))
        scale = float(state.get("scale", 1.0))
        self.resetTransform()
        self.scale(scale, scale)
        self.centerOn(float(center[0]), float(center[1]))
        self._shared_view_state = {"center": (float(center[0]), float(center[1])), "scale": scale}
        if emit:
            self.view_state_changed.emit(self.view_state())

    def keyPressEvent(self, event: QKeyEvent) -> None:
        scene = self.scene()
        if isinstance(scene, MapEditScene):
            scene.keyPressEvent(event)
            if event.isAccepted():
                return
        super().keyPressEvent(event)

    def reset_view(self) -> None:
        self.resetTransform()
        if self.scene() is not None:
            self.fitInView(
                self.scene().sceneRect(),
                Qt.AspectRatioMode.KeepAspectRatio,
            )
    view_state_changed = Signal(dict)
    cursor_position_changed = Signal(tuple)

File: libs/ui_pages_mapedit/src/map_edit_view.cpp
#include "pwb/ui_pages_mapedit/map_edit_view.hpp"

#include "pwb/ui_pages_mapedit/map_edit_scene.hpp"
#include "pwb/ui_shell/style_registry.hpp"

#include <QKeyEvent>
#include <QMouseEvent>
#include <QPainter>
#include <QTimer>
#include <QWheelEvent>

#include <cmath>

namespace pwb::ui_pages_mapedit {
namespace {

// _view_qss(): QGraphicsView#MapEditView themed sheet.
QString view_qss() {
    const auto& p = ui_shell::style_palette();
    const auto get = [&p](const char* k) {
        const auto it = p.find(k);
        return it != p.end() ? QString::fromStdString(it->second) : QString();
    };
    return QStringLiteral(
               "QGraphicsView#MapEditView { background: %1;"
               " border: 1px solid %2; border-radius: %3px; }")
        .arg(get("BG_SEARCH"), get("BORDER"), get("RADIUS_CARD"));
}

// _same_view_state — integer scrollbars round the center, so equality is
// tolerant (abs 1.5 on center, rel 1e-9 on scale).
bool same_view_state(const QVariantMap& a, const QVariantMap& b) {
    const auto center_of = [](const QVariantMap& m) {
        const QVariant c = m.value(QStringLiteral("center"));
        const auto l = c.toList();
        return l.size() >= 2
                   ? std::pair{l[0].toDouble(), l[1].toDouble()}
                   : std::pair{0.0, 0.0};
    };
    const auto [ax, ay] = center_of(a);
    const auto [bx, by] = center_of(b);
    const auto isclose = [](double x, double y, double abs_tol) {
        return std::fabs(x - y) <= abs_tol;
    };
    const double as = a.value(QStringLiteral("scale"), 1.0).toDouble();
    const double bs = b.value(QStringLiteral("scale"), 1.0).toDouble();
    // math.isclose(rel_tol=1e-9, abs_tol=0): tol scales with the larger
    // magnitude — no absolute floor.
    const double scale_tol =
        1e-9 * std::max(std::fabs(as), std::fabs(bs));
    return isclose(ax, bx, 1.5) && isclose(ay, by, 1.5) &&
           std::fabs(as - bs) <= scale_tol;
}

}  // namespace

MapEditView::MapEditView(QWidget* parent) : QGraphicsView(parent) {
    setObjectName(QStringLiteral("MapEditView"));
    // E1：动态注册（style.bind）——主题切换时按当前 palette 重渲染。
    ui_shell::style_bind(this, &view_qss);
    setRenderHint(QPainter::Antialiasing, true);
    setDragMode(QGraphicsView::NoDrag);
    setTransformationAnchor(QGraphicsView::AnchorUnderMouse);
    setResizeAnchor(QGraphicsView::AnchorUnderMouse);
    setViewportUpdateMode(QGraphicsView::SmartViewportUpdate);
    setFocusPolicy(Qt::StrongFocus);
    setMouseTracking(true);
    shared_view_state_ = {
        {QStringLiteral("center"), QVariantList{0.0, 0.0}},
        {QStringLiteral("scale"), 1.0},
    };
    // Navigation display LOD state (display-only; restored after idle).
    nav_lod_timer_ = new QTimer(this);
    nav_lod_timer_->setSingleShot(true);
    nav_lod_timer_->setInterval(kNavLodIdleMs);
    connect(nav_lod_timer_, &QTimer::timeout, this,
            [this]() { end_navigation_lod(); });

    setScene(new MapEditScene(this));
}

MapEditScene* MapEditView::edit_scene() const {
    return dynamic_cast<MapEditScene*>(scene());
}

// --- navigation display LOD -------------------------------------------------

void MapEditView::begin_navigation_lod() {
    // Enter low-detail mode for wheel/pan; an idle timer restores detail.
    if (!nav_lod_active_) {
        nav_lod_active_ = true;
        setRenderHint(QPainter::Antialiasing, false);
        setOptimizationFlags(QGraphicsView::DontAdjustForAntialiasing |
                             QGraphicsView::DontSavePainterState);
        if (auto* edit = edit_scene()) {
            edit->set_navigation_lod(true);
        }
    }
    nav_lod_timer_->start();  // restart the idle countdown
}

void MapEditView::end_navigation_lod() {
    if (!nav_lod_active_) {
        return;
    }
    nav_lod_active_ = false;
    nav_lod_timer_->stop();
    setRenderHint(QPainter::Antialiasing, true);
    setOptimizationFlags(QGraphicsView::OptimizationFlags());
    if (auto* edit = edit_scene()) {
        edit->set_navigation_lod(false);
    }
    viewport()->update();
    // Pan only reaches the view state here (wheel emits directly in
    // wheelEvent). Publish the settled state, but skip when unchanged so
    // apply_view_state(emit=false) and idle LOD cycles never echo.
    const QVariantMap new_state = read_view_state();
    if (!same_view_state(new_state, shared_view_state_)) {
        shared_view_state_ = new_state;
        emit view_state_changed(view_state());
    }
}

// --- events -----------------------------------------------------------------

void MapEditView::wheelEvent(QWheelEvent* event) {
    // Zoom with mouse wheel (no geometry change); low-detail while scrolling.
    const int delta = event->angleDelta().y();
    if (delta == 0) {
        event->ignore();
        return;
    }
    begin_navigation_lod();
    const double factor = delta > 0 ? 1.15 : 1.0 / 1.15;
    scale(factor, factor);
    shared_view_state_ = read_view_state();
    emit view_state_changed(view_state());
    event->accept();
}

void MapEditView::mouseMoveEvent(QMouseEvent* event) {
    // Publish the cursor in scene coordinates (project CRS).
    const QPointF pos = mapToScene(event->position().toPoint());
    emit cursor_position_changed(pos.x(), pos.y());
    QGraphicsView::mouseMoveEvent(event);
}

void MapEditView::scrollContentsBy(int dx, int dy) {
    // Any pan (scrollbars, keyboard, centerOn) engages navigation LOD.
    if (dx || dy) {
        begin_navigation_lod();
    }
    QGraphicsView::scrollContentsBy(dx, dy);
}

QVariantMap MapEditView::read_view_state() const {
    const QPointF center = mapToScene(viewport()->rect().center());
    return {
        {QStringLiteral("center"),
         QVariantList{center.x(), center.y()}},
        {QStringLiteral("scale"), transform().m11()},
    };
}

void MapEditView::apply_view_state(const QVariantMap& state,
                                   bool emit_change) {
    const QVariant c = state.value(QStringLiteral("center"));
    const auto l = c.toList();
    const double cx = l.size() >= 1 ? l[0].toDouble() : 0.0;
    const double cy = l.size() >= 2 ? l[1].toDouble() : 0.0;
    const double s = state.value(QStringLiteral("scale"), 1.0).toDouble();
    resetTransform();
    scale(s, s);
    centerOn(cx, cy);
    shared_view_state_ = {
        {QStringLiteral("center"), QVariantList{cx, cy}},
        {QStringLiteral("scale"), s},
    };
    if (emit_change) {
        emit view_state_changed(view_state());
    }
}

void MapEditView::reset_view() {
    resetTransform();
    if (scene() != nullptr) {
        fitInView(scene()->sceneRect(), Qt::KeepAspectRatio);
    }
}

void MapEditView::keyPressEvent(QKeyEvent* event) {
    if (auto* edit = edit_scene()) {
        edit->key_press(event);
        if (event->isAccepted()) {
            return;
        }
    }
    QGraphicsView::keyPressEvent(event);
}

}  // namespace pwb::ui_pages_mapedit
